### DISCUSSION TITLE
goyek vs Mage

### DISCUSSION THREAD

**Discussion by pellared** [dagger/dagger#3601] [Ideas]
Hello 👋 

Have you been considering using [goyek](https://github.com/goyek/goyek) instead of [Mage](https://github.com/magefile/mage)?

If so, could you share why have you chosen Mage over goyek? Otherwise, could you take a look and share your feedback if it would even consider using it if you were aware of it?

Feel free to ask any questions or share any feedback. You can also delete the discussion if you think it is a waste of your time 😉

**Comment by marcosnils** [comment]
:wave: 

AFAIK we haven't discussed using anything else instead of `mage` as the entrypoint for the Go SDK. In any case, `mage` is just a convenience and dagger should be able to play nicely with either framework / library.

> If so, could you share why have you chosen Mage over goyek? Otherwise, could you take a look and share your feedback if it would even consider using it if you were aware of it?

I see you're the creator / main contributor of `goyek`. Let me turn this question around. Is there  anything in `goyek` particularly that would benefit the dagger users over `mage` or anything else?  I've seen the alternatives section in goyek's README file but I don't feel that those points about mage have been a pain for us. 

In favor of mage, it's well known in the Go ecosystem and  an active project with certain degree of contributions coming in. We haven't seen or found anything similar that in our opinion provides a better experience to our users. If there's a reason to switch to something better, happy to discuss

**Reply by pellared** [reply]
TBH, I do not see any "deal-breakers". I think it is a lot a matter of preference, scale and trade-offs. I created this issue mainly to gather feedback. I do not insist (nor recommend) changing mage to goyek 😉 

Thanks a lot for your reply.

**Reply by sheldonhull** [reply]
@pellared has done a fantastic job on Goyek and we've had some great discussion on this in the past. I think it adheres to a more "pure" Go form that is very similar in style to Go tests. The catch is because it does less "magic", it results in effort, albeit not a lot, to generate reusable shared libraries to use across multiple repos. Definitely feels like more idiomatic Go. ⚡ 

The biggest difference for me personally was that Mage does some "magic" in discovering tasks via AST and the function signatures and has limited flag argument support. In contrast, Goyek does a fantastic job of being explicit and clear Go code with no magic autodiscovery, but this makes it harder to "auto-discover" remote packages as actual commands without adding the plumbing to connect those package commands into the goyek main func first. It's not bad, one might argue more idiomatic, but it is a big difference.

This "magic" from the Mage binary is really the plumbing that lets creation of tasks with Mage be so quick as it's bridging the gap from running `go run mage.go taskname` to `mage taskname` and importing and discovering functions based on that signature.

My personal take, having used Goyek in the past, is it's really cool where it's going and I'm glad to see more Go tools for this. 

I _prefer_ Mage right now, though it has shortcomings that Goyek solves, simply because: 

- well known with very stable api (i was earlier tester of Goyek and it wasn't yet stable with the api, which I'm sure has changed).
- while not "idiomatic" to Go, the auto-discover and import functions based on a package import makes it easier to share and reuse tasks without having extra registration effort.

Hopefully that's useful feedback both for @pellared and dagger folks.

If I had the time I'd be using both as @pellared has put a lot of great thought into it and I'm sure much of what makes Goyek unique would probably be something that Mage might have done differently if a version 2 later one had been possible. 

👍🏻 Cheers!

**Reply by pellared** [reply]
> import functions based on a package import makes it easier to share and reuse tasks without having extra registration effort.

It is still possible with `goyek`. Yet it is not forced and it is not [well] documented 😉 One can define tasks using `goyek.Define` in shared packages and then they can be imported using the `import _ package/path/syntax`.

**Comment by shykes** [comment]
Just to clarify @pellared , Dagger does not actually use Mage in its Go SDK, nor does it require you to use Mage.

As users of Dagger ourselves, we (the Dagger team) happen to use Dagger with Mage, but that’s just a personal choice. We could just as easily write our own `main` function, or use another tool.

I didn’t know about goyek, but would love to see an example of using Dagger and goyek together, if the community finds it helpful perhaps we could make it a guide in our docs. Is this something you’d be interested in contributing?

**Reply by pellared** [reply]
I can create a PR which ports usage of mage to goyek (of course I will leave the `magefile.go`). From that point, I think you can decide if it is worth any further investment. Does it sound reasonable?

**Reply by pellared** [reply]
I did it to keep the momentum https://github.com/dagger/dagger/pull/3605
